src/Native/Unix/System.Native/pal_io.c

```diff
@@ -340,22 +346,25 @@ int32_t SystemNative_GetReadDirRBufferSize(void)
 // If the platform supports readdir_r, the caller provides a buffer into which the data is read.
 // If the platform uses readdir, the caller must ensure no calls are made to readdir/closedir since those will invalidate
 // the current dirent. We assume the platform supports concurrent readdir calls to different DIRs.
-int32_t SystemNative_ReadDirR(DIR* dir, void* buffer, int32_t bufferSize, struct DirectoryEntry* outputEntry)
+int32_t SystemNative_ReadDirR(DIR* dir, uint8_t* buffer, int32_t bufferSize, struct DirectoryEntry* outputEntry)
 {
     assert(dir != NULL);
     assert(outputEntry != NULL);
 
 #if HAVE_READDIR_R
     assert(buffer != NULL);
 
-    if (bufferSize < (int32_t)sizeof(struct dirent))
+    // align to dirent
+    struct dirent* entry = (struct dirent*)((size_t)(buffer + dirent_alignment - 1) & ~(dirent_alignment - 1));
+
+    // check there is dirent size available at entry
+    if ((buffer + bufferSize) < (uint8_t*)(entry + 1))
```

Comment:
How does this check that there's space available for dirent?

Reply:
the lhs is the address after buffer, the rhs is the address after entry. `entry + 1` takes adds `sizeof(struct dirent)`. I'll change the rhs to `(uint8_t*)entry + sizeof(struct dirent)` to make it more explicit.

Reply:
Aha.  I missed that entry was declared as `struct dirent*` rather than `uint8_t*`.  Yeah, making it more explicit would be helpful.